File: mars_rover/mars_rover/mars_rover.py
from dataclasses import dataclass


@dataclass(init=True)
class Position:
    x: int
    y: int

    def __str__(self):
        return f"{self.x}:{self.y}"


class Direction:
    _directions = [
        'N', 'E', 'S', 'W'
    ]

    def __init__(self, facing: str):
        self._current_direction = self._directions.index(facing)

    def turn_right(self):
        self._current_direction += 1
        if self._current_direction >= len(self._directions):
            self._current_direction = 0

    def turn_left(self):
        self._current_direction -= 1
        if self._current_direction < 0:
            self._current_direction = len(self._directions) - 1

    def facing(self):
        return self._directions[self._current_direction]


class MarsRover:
    GRID_SIZE = 10

    _position: Position

    def __init__(self):
        self.direction = Direction(facing='N')
        self._position = Position(x=0, y=0)
        self._facing = 'N'

    def execute(self, command):
        if not command:
            return self.facing()

        for turn in command:
            if turn == 'R':
                self.direction.turn_right()
            if turn == 'L':
                self.direction.turn_left()
            if turn == 'M':
                self._move(self.direction.facing())

        return self.facing()

    def _move(self, facing: str):
        if facing == 'E':
            self._position.x += 1
        if facing == 'W':
            self._position.x -= 1
        if facing == 'S':
            self._position.y -= 1
        elif facing == 'N':
            self._position.y += 1

        if self._position.y % self.GRID_SIZE == 0:
            self._position.y = 0

        if self._position.x % self.GRID_SIZE == 0:
            self._position.x = 0

    def facing(self):
        return f'{self._position}:{self.direction.facing()}'

File: mars-rover-java/app/src/main/java/com/github/marsrover/MarsRover.java
package com.github.marsrover;

public class MarsRover {
    private Direction currentDirection;
    private Coordinate coordinates;
    private final Grid grid;

    public MarsRover() {
        currentDirection = Direction.NORTH;
        coordinates = new Coordinate(0, 0);
        grid = new Grid();
    }

    public String execute(String command) {
        for (char cmd : command.toCharArray()) {
            String singleCommand = String.valueOf(cmd);
            if (singleCommand.equals("R")) {
                currentDirection = currentDirection.turnRight();
            }

            if (singleCommand.equals("L")) {
                currentDirection = currentDirection.turnLeft();
            }

            if (singleCommand.equals("M")) {
                coordinates = grid.next(coordinates, currentDirection);
            }
        }

        return facing();
    }

    private String facing() {
        return String.format(
                "%d:%d:%s",
                coordinates.x(),
                coordinates.y(),
                currentDirection.value()
        );
    }
}
